caffe2/serialize/intermediate_model.cc

```diff
@@ -0,0 +1,334 @@
+    submodules_.at(i).dump(sub_def);
+  }
+
+  for (int i = 0; i < methods_.size(); ++i) {
+    module_def->add_methods();
+    torch::MethodDef* method_def = module_def->mutable_methods(module_def->methods_size() - 1);
+    methods_.at(i).dump(method_def);
+  }
+}
+
+void IntermediateModel::update(torch::ModelDef* model_def,
+    std::unordered_map<uint64_t, std::shared_ptr<SharedData>>* id_data,
+    DeserializeMode mode) {
+  AT_ASSERTM(model_def->has_name(), "name is required for ModelDef.");
+  name_ = model_def->name();
+  AT_ASSERTM(model_def->has_producer_name(), "producer_name is required for ModelDef.");
+  producerName_ = model_def->producer_name();
+  producerVersion_ = model_def->producer_version();
+  AT_ASSERTM(model_def->has_proto_version(), "proto_version is required for ModelDef.");
+  protoVersion_ = model_def->proto_version();
+  AT_ASSERTM(model_def->has_main_module(), "main_module is required for ModelDef.");
+  mainModule_.update(model_def->mutable_main_module(), id_data, mode);
+}
+
+void IntermediateModel::dump(torch::ModelDef* model_def) {
+  model_def->set_name(name_);
+  model_def->set_producer_name(producerName_);
+  model_def->set_producer_version(producerVersion_);
+  model_def->set_proto_version(protoVersion_);
+  mainModule_.dump(model_def->mutable_main_module());
+}
+
+void serializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileWriter* writer) {
+  std::unordered_map<void*, uint64_t> data_id;
+  std::stack<IntermediateModule*> imodules;
+  imodules.push(imodel->mutableMainModule());
+  while (!imodules.empty()) {
+    IntermediateModule* m = imodules.top();
+    imodules.pop();
+    auto* params = m->mutableParameters();
+    for (int i = 0; i < params->size(); ++i) {
+      std::shared_ptr<SharedData> dataptr = params->at(i).mutableTensor()->data();
+      void* data = dataptr->dataPtr.get();
+      size_t size = dataptr->size;
+      auto it = data_id.find(data);
+      if (it != data_id.end()) {
+        dataptr->recordId = it->second;
+      } else {
+        uint64_t id = writer->writeRecord(data, size);
+        dataptr->recordId = id;
+        data_id[data] = id;
+      }
+    }
+
+    auto* subms = m->mutableSubmodules();
+    for (int i = 0; i < subms->size(); ++i) {
+      imodules.push(&subms->at(i));
+    }
+  }
+
+  torch::ModelDef model_def;
+  imodel->dump(&model_def);
+  std::string buffer;
+  model_def.SerializeToString(&buffer);
+  writer->writeRecord(buffer.c_str(), buffer.size());
+}
+
+void serializeIntermediateModel(IntermediateModel* imodel, const std::string& filename) {
+  torch::jit::PyTorchFileWriter writer(filename);
+  serializeIntermediateModel(imodel, &writer);
+  writer.writeEndOfFile();
+}
+
+void deserializeIntermediateModel(IntermediateModel* imodel,
+    torch::jit::PyTorchFileReader* reader, DeserializeMode mode) {
+  std::unordered_map<uint64_t, std::shared_ptr<SharedData>> id_data;
+  if (mode == DeserializeMode::LAZY) {
```

Comment:
LAZY/EAGER are too easy to mix up with run modes. How about LOAD_TENSOR_DATA vs HEADER_ONLY?

Reply:
Sounds great!